temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+            shared_manager: Optional[SharedStateManager] = None
+            if not isinstance(input.executor, concurrent.futures.ThreadPoolExecutor):
+                # Should always be present in worker, pre-checked on init
+                shared_manager = input._worker._config["shared_state_manager"]
+                assert shared_manager
+                heartbeat = shared_manager.register_heartbeater(
+                    info.task_token, ctx.heartbeat
+                )
+
+            try:
+                return await asyncio.get_running_loop().run_in_executor(
+                    input.executor,
+                    _execute_sync_activity,
+                    info,
+                    heartbeat,
+                    # Only thread event, this may cross a process boundary
+                    input._cancelled_event.thread_event,
+                    input._worker_shutdown_event.thread_event,
+                    input.fn,
+                    *input.args,
+                )
+            finally:
+                if shared_manager:
+                    shared_manager.unregister_heartbeater(info.task_token)
+
+        # Otherwise for async activity, just run
+        return await input.fn(*input.args)
+
+
+# This has to be defined at the top-level to be picklable for process executors
+def _execute_sync_activity(
+    info: temporalio.activity.Info,
+    heartbeat: Union[Callable[..., None], SharedHeartbeatSender],
+    cancelled_event: threading.Event,
+    worker_shutdown_event: threading.Event,
+    fn: Callable[..., Any],
+    *args: Any,
+) -> Any:
+    heartbeat_fn: Callable[..., None]
+    if isinstance(heartbeat, SharedHeartbeatSender):
+        # To make mypy happy
+        heartbeat_sender = heartbeat
+        heartbeat_fn = lambda *details: heartbeat_sender.send_heartbeat(
+            info.task_token, *details
+        )
+    else:
+        heartbeat_fn = heartbeat
+    temporalio.activity._Context.set(
+        temporalio.activity._Context(
+            info=lambda: info,
+            heartbeat=heartbeat_fn,
+            cancelled_event=temporalio.activity._CompositeEvent(
+                thread_event=cancelled_event, async_event=None
+            ),
+            worker_shutdown_event=temporalio.activity._CompositeEvent(
+                thread_event=worker_shutdown_event, async_event=None
+            ),
+        )
+    )
+    return fn(*args)
+
+
+class _ActivityOutboundImpl(ActivityOutboundInterceptor):
+    def __init__(self, worker: Worker, info: temporalio.activity.Info) -> None:
+        # We are intentionally not calling the base class's __init__ here
+        self._worker = worker
+        self._info = info
+
+    def info(self) -> temporalio.activity.Info:
+        return self._info
+
+    def heartbeat(self, *details: Any) -> None:
+        info = temporalio.activity.info()
+        self._worker._heartbeat_activity(info.task_token, *details)
+
+
+def _proto_maybe_timedelta(
```

Comment:
No, there are no cases in the client (yet) of using a proto duration. Also, this is just a thin wrapper to support 0 durations, which I don't think we want even with retry policy. Otherwise, the `To`/`From` `Timedelta/Datetime` that come w/ Python proto are good enough for common use.

Reply:
Inlined except for the date one where I need the TZ